### PR TITLE
chore: add more ts type definitions

### PR DESCRIPTION
**Pull request by AndreyBykov**
Second portion of TS type definitions, jsdoc cleaned up + updated the tests where necessary.

Main question I have here is related to the dataset. There's `DatasetDataOptions` interface, which has all params like `limit`, `skipHidden`, `fields` and so on. And there are methods such as `forEach` and some others. `forEach` has 3 options exposed (desc, fields, unwind). While inside of the method there are checks for `format`, `limit`, maybe some others. So I could basically use the `DatasetDataOptions` interface there, but in order to hide some params, I am using `DatasetIteratorOptions extends DatasetDataOptions` with some params overridden with `@internal`. Is it the correct way to do it or what's the proper way in such cases?

P.S.: sry for the delay, will start with the remaining part right away.

### CODE REVIEW

**Comment by vladfrangu** [comment]
> So I could basically use the DatasetDataOptions interface there, but in order to hide some params, I am using DatasetIteratorOptions extends DatasetDataOptions with some params overridden with @internal.

Ideally you could make a base interface with common props, then extend it as needed, or use Omit to remove properties that aren't needed

File: packages/apify/src/storages/key_value_store.ts [hunk 1]
```diff
@@ -205,18 +205,15 @@ export class KeyValueStore {
...
-    // TODO: type options
-    async setValue<T>(key: string, value: T | null, options: Dictionary = {}): Promise<void> {
+    async setValue<T>(key: string, value: T | null, options: RecordOptions = {}): Promise<void> {
         ow(key, 'key', ow.string.nonEmpty);
         ow(key, ow.string.validate((k) => ({
             validator: ow.isValid(k, ow.string.matches(KEY_VALUE_STORE_KEY_REGEX)),
             message: 'The "key" argument must be at most 256 characters long and only contain the following characters: a-zA-Z0-9!-_.\'()',
         })));
-        if (options.contentType && !ow.isValid(value, ow.any(ow.string, ow.buffer))) {
+        if (options!.contentType && !ow.isValid(value, ow.any(ow.string, ow.buffer))) {
```
Comment (vladfrangu) [inline packages/apify/src/storages/key_value_store.ts:216]: is the `!` needed?

File: packages/apify/src/storages/key_value_store.ts [hunk 2]
```diff
@@ -436,11 +405,9 @@ export async function getValue<T extends Dictionary | string | Buffer>(key: stri
...
- * @param [options]
- * @param {string} [options.contentType]
- *   Specifies a custom MIME content type of the record.
+ * @param [options] Record options.
  */
-export async function setValue<T>(key: string, value: T, options?: unknown): Promise<void> {
+export async function setValue<T>(key: string, value: T, options: RecordOptions = {}): Promise<void> {
     const store = await openKeyValueStore();
 
     // @ts-ignore options need to be typed
```
Comment (vladfrangu) [inline packages/apify/src/storages/key_value_store.ts:413]: This isn't needed anymore, right?

File: packages/apify/src/storages/key_value_store.ts [hunk 3]
```diff
@@ -498,3 +465,17 @@ export interface KeyValueStoreOptions {
     client: ApifyClient | ApifyStorageLocal;
     isLocal?: boolean;
 }
+
+interface RecordOptions {
```
Comment (vladfrangu) [inline packages/apify/src/storages/key_value_store.ts:469]: Please export the option interfaces

**Review by vladfrangu — CHANGES_REQUESTED** [review]
Awesome start! Just a few small nits from me

**Comment by AndreyBykov** [comment]
@vladfrangu but that's the thing, there's `getData` call (which takes all options) withing `forEach` (here we expose only 3 props). So I guess they should not be omitted at all, but rather be internal in this case...

File: packages/apify/src/storages/request_queue.ts [hunk 4]
```diff
@@ -596,7 +593,7 @@ export class RequestQueue {
      * Adds a request straight to the queueHeadDict, to improve performance.
      * @private
      */
-    _maybeAddRequestToQueueHead(requestId, forefront) {
+    _maybeAddRequestToQueueHead(requestId: string, forefront: boolean): void {
```
Comment (B4nan) [inline packages/apify/src/storages/request_queue.ts:596]: is this used from outside? if not, make it `private` on TS level

File: packages/apify/src/storages/request_queue.ts [hunk 5]
```diff
@@ -676,9 +669,10 @@ export class RequestQueue {
      *   ID or name of the request queue to be opened. If `null` or `undefined`,
      *   the function returns the default request queue associated with the actor run.
      * @param [options]
-     * @param [options.forceCloud=false]
+     * @param [options.forceCloud]
      *   If set to `true` then the function uses cloud storage usage even if the `APIFY_LOCAL_STORAGE_DIR`
      *   environment variable is set. This way it is possible to combine local and cloud storage.
+     * @default false
      */
     static async open(queueIdOrName?: string | null, options: { forceCloud?: boolean } = {}): Promise<RequestQueue> {
```
Comment (B4nan) [inline packages/apify/src/storages/request_queue.ts:677]: let's make this an interface too, i don't think the API docs can handle the `options.forceCloud` jsdoc

File: test/storages/request_queue.test.ts [hunk 6]
```diff
@@ -512,13 +513,13 @@ describe('RequestQueue remote', () => {
         const updateRequestMock = jest.spyOn(queue.client, 'updateRequest');
         updateRequestMock.mockResolvedValueOnce({ requestId: 'b', wasAlreadyHandled: false, wasAlreadyPresent: true });
 
-        await queue.markRequestHandled(requestBWithId);
+        await queue.markRequestHandled(requestBWithId as ApifyRequest);
```
Comment (B4nan) [inline test/storages/request_queue.test.ts:516]: let's move the type assertion to the place where the variable is defined

**Review by B4nan — CHANGES_REQUESTED** [review]
looking good, few more nits from my end

File: packages/apify/src/storages/dataset.ts [hunk 7]
```diff
@@ -85,6 +85,81 @@ export function chunkBySize(items: string[], limitBytes: number): string[] {
...
+    /**
+     * If `true` then the function doesn't return empty items.
+     * Note that in this case the returned number of items might be lower than limit parameter and pagination must be done using the `limit` value.
+     * @default false
+     */
+    skipEmpty?: boolean;
+}
+
+interface DatasetIteratorOptions extends DatasetDataOptions {
+    /** @internal */
```
[the remark this answers is not in the document]
Comment (AndreyBykov) [reply packages/apify/src/storages/dataset.ts:141]: But looking at the current docs - only 3 properties are exposed in `.forEach` and `.map` methods. I thought if I would just use the main interface (without hidden/internal props) - we would show all the props for these methods? Is it kinda ok?

[the remark this answers is not in the document]
Comment (B4nan) [reply packages/apify/src/storages/dataset.ts:141]:
```ts
export interface DatasetIteratorOptions extends Omit<DatasetDataOptions, 'offset' | 'limit' ...> {
```

File: packages/apify/src/storages/key_value_store.ts [hunk 8]
```diff
@@ -281,26 +273,16 @@ export class KeyValueStore {
...
-     * @param {Record<string, any>} [options]
-     * @param {number} [index=0]
-     * @return {Promise<Promise<void> | undefined>}
-     * @private
-     */
-    private async _forEachKey(iteratee: KeyConsumer, options: Dictionary<string | undefined> = {}, index = 0): Promise<void> {
-        const { exclusiveStartKey } = options;
+    /** @private */
+    private async _forEachKey(iteratee: KeyConsumer, options: KeyValueStoreIteratorOptions = {}, index = 0): Promise<void> {
+        const { exclusiveStartKey } = options!;
```
[the remark this answers is not in the document]
Comment (AndreyBykov) [reply packages/apify/src/storages/key_value_store.ts:285]: yeah, I removed the default value at the beginning - TS was not happy, then I realized it should be there and forgot to remove the `!`

File: packages/apify/src/storages/request_queue.ts [hunk 9]
```diff
@@ -668,13 +672,10 @@ export class RequestQueue {
      * @param [queueIdOrName]
      *   ID or name of the request queue to be opened. If `null` or `undefined`,
      *   the function returns the default request queue associated with the actor run.
-     * @param [options]
-     * @param [options.forceCloud]
-     *   If set to `true` then the function uses cloud storage usage even if the `APIFY_LOCAL_STORAGE_DIR`
-     *   environment variable is set. This way it is possible to combine local and cloud storage.
+     * @param [options] Open Request Queue options.
      * @default false
```
Comment (B4nan) [inline packages/apify/src/storages/request_queue.ts:676]: this belongs to the `forceCloud`, right?

**Review by B4nan — APPROVED** [review]

File: packages/apify/src/storages/dataset.ts [hunk 10]
```diff
@@ -85,6 +85,81 @@ export function chunkBySize(items: string[], limitBytes: number): string[] {
...
+     * If `true` then the function doesn't return empty items.
+     * Note that in this case the returned number of items might be lower than limit parameter and pagination must be done using the `limit` value.
+     * @default false
+     */
+    skipEmpty?: boolean;
+}
+
+export interface DatasetIteratorOptions extends Omit<DatasetDataOptions, 'offset' | 'limit' | 'clean' | 'skipHidden' | 'skipEmpty'> {
+    /** @internal */
+    offset?: number;
```
[the remark this answers is not in the document]
Comment (AndreyBykov) [reply packages/apify/src/storages/dataset.ts:142]: Yeah, we had a call with Martin. It's more about docs. These fields are further used inside of the method, but we do not expose them, so not to have any clashes (public/internal) - I omit and re-define them there with @internal.

File: packages/apify/src/storages/request_queue.ts [hunk 11]
```diff
@@ -500,21 +504,19 @@ export class RequestQueue {
...
-     * @param {boolean} [ensureConsistency=false] If true then query for queue head is retried until queueModifiedAt
+     * @param [ensureConsistency] If true then query for queue head is retried until queueModifiedAt
      *   is older than queryStartedAt by at least API_PROCESSED_REQUESTS_DELAY_MILLIS to ensure that queue
      *   head is consistent.
-     * @param {number} [limit] How many queue head items will be fetched.
-     * @param {number} [iteration] Used when this function is called recursively to limit the recursion.
+     * @default false
+     * @param [limit] How many queue head items will be fetched.
+     * @param [iteration] Used when this function is called recursively to limit the recursion.
      * @return {Promise<boolean>} Indicates if queue head is consistent (true) or inconsistent (false).
```
[the remark this answers is not in the document]
Comment (AndreyBykov) [reply packages/apify/src/storages/request_queue.ts:513]: Yeah, think so. Will go through everything once again, thanks for noticing!

**Comment by AndreyBykov** [comment]
I went through all folders once again, could you guys check the last commit?
@vladfrangu - removed the return types.
@B4nan - cleaned up the @internal for protected/private methods.
Wanted to include both in this PR since we discussed it before in comments.

**Review by B4nan — APPROVED** [review]

File: packages/apify/src/session_pool/session.ts [hunk 12]
```diff
@@ -222,11 +218,13 @@ export class Session {
...
     /**
      * Gets session state for persistence in KeyValueStore.
-     * @return {SessionState} represents session internal state.
+     * @returns Represents session internal state.
      */
-    getState() {
+    getState(): SessionState {
         return {
             id: this.id,
+            // @ts-expect-error Type 'Serialized' is missing the following properties from type 'CookieJar':
```
Comment (vladfrangu) [inline packages/apify/src/session_pool/session.ts:226]: This doesn't seem right, is cookieJar meant to be an object or an actual instance?

**Review by vladfrangu — APPROVED** [review]
LGTM, just one question for now

File: packages/apify/src/session_pool/session.ts (hunk 12 above)
Comment (AndreyBykov) [reply packages/apify/src/session_pool/session.ts:226]: @vladfrangu I think here it's supposed to be the serialized. Session in memory keeps the actual instance of CookieJar. But this method gets the session state to be persisted in KeyValueStore and some fields are omitted (e.g. setCookie, setCookieSync methods) - thus the error.

**Pull request merged**